crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1340,21 +1365,77 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Resolves the length value of a type.
+    ///
+    /// This is used to determine the value that would be returned
+    /// when `len(x)` is called on an object `x`.
+    fn len(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        match self {
+            Type::BytesLiteral(bytes) => return bytes.len(db).try_into().ok(),
+            Type::StringLiteral(string) => {
+                return string.as_str(db).chars().count().try_into().ok()
+            }
```

Comment:
I think you've correctly identified here that `StringLiteralType::len` would be subtly incorrect for your use case here. However, `StringLiteralType::len` is only used in one place currently, and I think its implementation is subtly incorrect for _that_ use case as well. It's used here, in our implementation of unpacked assignments:

https://github.com/astral-sh/ruff/blob/d9cbf2fe44e3f13b28295377a41ae541cab6b54f/crates/red_knot_python_semantic/src/types/unpacker.rs#L90-L101

But according to the [Rust docs for `str`](https://doc.rust-lang.org/std/primitive.str.html#method.len), `"ƒoo".len() == 4` in Rust (because of the fancy Unicode `f`). That gives us the wrong result for Python unpacking:

```pycon
>>> a, b, c = "ƒoo"
>>> a
'ƒ'
>>> b
'o'
>>> c
'o'
```

Therefore, I think rather than adding a separate `StringLiteralType::as_str()`, we should simply change `StringLiteralType::len` to count the characters rather than the bytes (and add a comment about why). Per Micha's earlier comment, it might be best to rename it to `python_len()`, so that we don't get confused about how it differs from the way that Rust counts the length of the string